eth/filters,ethclient,node: install newSideHeads subscription

meowsbits opened the pull request:
The newSideHeads subscription work very similarly to
the newHeads subscription; instead, non-canonical blocks are channeled.

Resolves #287

meowsbits committed:
eth/filters,ethclient,node: install newSideHeads subscription

The newSideHeads subscription work very similarly to
the newHeads subscription; instead, non-canonical blocks are channeled.

meowsbits commented on eth/filters/filter_system.go:
```diff
@@ -399,6 +440,10 @@ func (es *EventSystem) lightFilterNewHead(newHeader *types.Header, callBack func
 }
 
+func (es *EventSystem) lightFilterNewSideHead(header *types.Header, callBack func(*types.Header, bool)) {
+	callBack(header, true)
```
This is noteworthy; logs from side chain blocks will always be marked `Removed: true`. `newHeads` handling of this uses `Removed` to indicate the same characteristic, but in that context is expected to _have been_ canonical, and is now _removed_ to the side. Although the reuse of this field (and vocabulary) is a little awkward, I think it's the best we can do without modifying the existing API data type.

meowsbits committed:
filters: (lint) goimports

Date: 2021-01-12 08:05:16-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr commented on eth/filters/api.go:
```diff
@@ -233,7 +266,37 @@ func (api *PublicFilterAPI) NewHeads(ctx context.Context) (*rpc.Subscription, er
 }
 
-// Logs creates a subscription that fires for all new log that match the given filter criteria.
+// NewHeads send a notification each time a new (header) block is appended to the chain.
```
This comment is wrong

meowsbits committed:
filters: (lint) fix comment

Resolves https://github.com/etclabscore/core-geth/pull/293/files/bf5bd1290fdb0ad59e8d4255100934d053be6e5d#r555805816

Date: 2021-01-12 09:54:37-06:00
Signed-off-by: meows <b5c6@protonmail.com>

ziogaschr reviewed (approved):
LGTM!

meowsbits committed:
ethclient: install missing eth_newSideBlockFilter method for rpc.discover test

Date: 2021-01-19 07:07:02-06:00
Signed-off-by: meows <b5c6@protonmail.com>
